core/domain/recommendations_validators.py

```diff
@@ -0,0 +1,111 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Validators for recommendation services in prod."""
+
+from __future__ import absolute_import  # pylint: disable=import-only-modules
+from __future__ import unicode_literals  # pylint: disable=import-only-modules
+
+from core.domain import base_model_validators
+from core.domain import recommendations_services
+from core.platform import models
+import python_utils
+
+(
+    exp_models, recommendations_models) = (
+        models.Registry.import_models([
+            models.NAMES.exploration,
+            models.NAMES.recommendations]))
+
+
+class ExplorationRecommendationsModelValidator(
+        base_model_validators.BaseModelValidator):
+    """Class for validating ExplorationRecommendationsModel."""
+
+    @classmethod
+    def _get_external_id_relationships(cls, item):
+        return [
+            base_model_validators.ExternalModelFetcherDetails(
+                'exploration_ids', exp_models.ExplorationModel,
+                [item.id] + item.recommended_exploration_ids)]
+
+    @classmethod
+    def _validate_item_id_not_in_recommended_exploration_ids(cls, item):
+        """Validate that model id is not present in recommended
+        exploration ids.
+
+        Args:
+            item: datastore_services.Model. ExplorationRecommendationsModel
+                to validate.
+        """
+
+        if item.id in item.recommended_exploration_ids:
+            cls._add_error(
+                'item exploration %s' % (
+                    base_model_validators.ERROR_CATEGORY_ID_CHECK),
+                'Entity id %s: The exploration id: %s for which the entity is '
+                'created is also present in the recommended exploration ids '
+                'for entity' % (item.id, item.id))
+
+    @classmethod
+    def _get_custom_validation_functions(cls):
+        return [cls._validate_item_id_not_in_recommended_exploration_ids]
+
+
+class TopicSimilaritiesModelValidator(base_model_validators.BaseModelValidator):
+    """Class for validating TopicSimilaritiesModel."""
+
+    @classmethod
+    def _get_model_id_regex(cls, unused_item):
+        # Valid id: topics.
+        return '^%s$' % recommendations_models.TOPIC_SIMILARITIES_ID
+
+    @classmethod
+    def _get_external_id_relationships(cls, item):
+        return []
+
+    @classmethod
+    def _validate_topic_similarities(cls, item):
+        """Validate the topic similarities to be symmetric and have real
+        values between 0.0 and 1.0.
+
+        Args:
+            item: datastore_services.Model. TopicSimilaritiesModel
+                to validate.
+        """
+
+        topics = list(item.content.keys())
+        data = '%s\n' % ','.join(topics)
+
+        for topic1 in topics:
+            similarity_list = []
+            for topic2 in item.content[topic1]:
```

Comment:
Rename topic1 and topic2 to something more meaningful? Like `topic` and `topic_to_compare` maybe? not sure of the context here.

Reply:
done.